File: store/store_products/products_app/mqtt.py
import logging
import uuid
import paho.mqtt.client as mqtt
from enum import Enum

import time
import uuid

import json


'''
--------------- HOST and PORT ---------------
'''
# Host
MQTT_HOST = "broker.hivemq.com"

# Port
MQTT_PORT = 1883

# Client id
MQTT_CLIENT_ID = str(uuid.uuid4())


'''
--------------- TOPICS ---------------
'''
# Get the read topic for the store
# Get the current timestamp in milliseconds
timestamp_ms = int(round(time.time() * 1000))
# Generate a random UUID
uid = str(uuid.uuid4())
# Create the topic name with the timestamp and UID appended
TOPIC_READ = "store-products/complete-payments-{}-{}".format(timestamp_ms, uid)


# Bank topic to send the payment request
TOPIC_WRITE = "tiago-bank/transactions"





'''
--------------- LOGIC ---------------
'''
def on_connect(client, userdata, flags, rc):
    print("Connected with result code " + str(rc))
    client.subscribe(TOPIC_READ)

def on_message(client, userdata, msg):
    # Handle the received message here
    payload = json.loads(msg.payload.decode('utf-8'))
    message = Message(**payload)
    
    from products_app.views import mark_as_payment_received
    logging.info(msg.topic + " " + str(msg.payload)+ "\n")
    mark_as_payment_received(None, message.orderId)


class Message:
    def __init__(self, ibanSender, ibanReceiver, amount, date, orderId, confirmationTopic):
        self.ibanSender = ibanSender
        self.ibanReceiver = ibanReceiver
        self.amount = amount
        self.date = date
        self.orderId = orderId
        self.confirmationTopic = confirmationTopic

    def __str__(self):
        return f"{self.ibanSender} {self.ibanReceiver} {self.amount} {self.date} {self.orderId} {self.confirmationTopic}"


client = mqtt.Client()
client.on_connect = on_connect
client.on_message = on_message
client.connect(MQTT_HOST, MQTT_PORT, 60)

File: publicitaki/pub-users/mqtt.js
const mqtt = require('mqtt');
const { createOrUpdateSubscriptionForStoreById } = require('./services/subscriptionService');

// TOPICS (store subscription's payment)
const TOPIC_WRITE = "tiago-bank/transactions"; /* Topic to pay store's subscription */
const TOPIC_READ = `Publicitaki-${Date.now()}-${Math.floor(Math.random() * 1000000)}`; /* Topic to receive the confirmation from the bank */

const client = mqtt.connect('tcp://broker.hivemq.com:1883');

client.on('connect', function () {
    console.log('Connected to MQTT broker');
    client.subscribe(TOPIC_READ);
});

client.on('message', function (topic, message) {
    handleMessage(message);
});

function handleMessage(message) {

    // NOTE: orderId = StoreId (because of store system)

    const json = JSON.parse(message);
    if (!json.orderId) {
        console.warn("Invalid MQTT message");
    }

    // create a new Date object for the current moment
    const currentDate = new Date();

    // add 1 month to the current date
    const oneMonthLater = new Date();
    oneMonthLater.setMonth(currentDate.getMonth() + 1);

    try {
        createOrUpdateSubscriptionForStoreById(oneMonthLater.toString(), 'true', json.orderId);
        console.log("Subscription activated for store = " + json.orderId);
    } catch (err) {
        console.error("Error updating subscription");
    }
}




module.exports = {
    client,
    TOPIC_WRITE,
    TOPIC_READ
}
